cuML/DEVELOPER_GUIDE.md

```diff
@@ -0,0 +1,119 @@
+# cuML developer guide
+This document summarizes rules and best practices for contributions to C++ component cuML of RAPIDS/cuml.
+
+## General
+Please start by reading [CONTRIBUTING.md](../CONTRIBUTING.md).
+
+## Thread safety
+cuML should be thread safe and its functions can be called from multiple host threads if they use different handles.
```

Comment:
I fully agree the wording should be clear and "should" is not a good choice. Here I have started with "should" because this started as a proposal for discussion. However since that it is a proposal is obvious from the context.